### ISSUE TITLE
Document the algorithm behind optuna

### ISSUE THREAD

**Issue by twolodzko** [optuna/optuna#835]
As a user/researcher I would like to easily be able to find information on what kind of algorithm stands behind optuna, to be able to make informed decision on using it in my work. Currently main page says only that it does "Bayesian optimization".

As a user I would like to see detailed documentation about the algorithm behind the `Study.optimize` method, when I check the documentation page for this method, so that I can understand what does the method do. Currently the documentation page only says that it "optimizes an objective function".

**Comment by yongzhuo** [comment]
same question

**Comment by hvy** [comment]
Thanks for addressing problems with the documentation. I'm wondering if you could elaborate on the "optimizer" that you are mentioning. You do not mean the samplers such as the TPE sampler? The reason I'm asking is because [the documentation of this sampler](https://optuna.readthedocs.io/en/latest/reference/samplers.html#optuna.samplers.TPESampler) contains references to the original paper. 

As for the implementation details, I agree that it could be of interest to certain users but it doesn't necessarily have to be a part of the documentation. Again, it'd help if you could provide some concrete examples.

**Comment by twolodzko** [comment]
@hvy most of the examples in the documentation take form of

```{python}
import optuna
import pandas

def objective(trial):
    x = trial.suggest_uniform('x', -1, 1)
    return x ** 2

study = optuna.create_study()
study.optimize(objective, n_trials=3)
```

it does not mention sampler (I missed the referred documentation page). Moreover, it does not say what exactly does `optimize` method do? The documentation for [`optimize`](https://optuna.readthedocs.io/en/latest/reference/study.html#optuna.study.Study.optimize) just says that is "optimizes". I'd expect to see some details on this documentation page.

**Comment by hvy** [comment]
Sorry for the delayed response, and thanks. I agree that the current documentation somewhat assumes that TPE would be the obvious choice for many and that details aren't of that much interest. At the same time, https://optuna.readthedocs.io/en/latest/reference/study.html#optuna.study.create_study will guide you to [the page listing all sampler](https://optuna.readthedocs.io/en/latest/reference/samplers.html#module-optuna.samplers) (although not exhaustive as e.g. the CMA-ES sampler is not listed here, but there is a graph explaining internals).

I definitely agree though that the documentation for optimize can be improved. Let me label this issue as contribution-welcome.

**Comment by twolodzko** [comment]
When I have more time to play around with it and go through some of the code, I'll consider contributing by improving the documentation.

**Comment by sile** [comment]
As a merged PR #940 addressed this issue, please let me close this.
